### PR TITLE
Gen2

### PR DESCRIPTION
**Pull request by alanvgreen**
Initial gen2 code: multiply accumulate blocks and some post-processing logic.

### CODE REVIEW

File: proj/hps_accel/gateware/gen2/macc.py [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+            ab = Signal.like(a * b)
+            m.d.sync += ab.eq(a * b)
+            products.append(ab)
+
+        # Pipeline cycle 1: accumulate
+        product_sum = Signal.like(tree_sum(products))
+        m.d.comb += product_sum.eq(tree_sum(products))
+        first_delayed = self._delay(m, self.input_first, 1)
+        accumulator = Signal(self._accumulator_shape)
+        base = Mux(first_delayed[0], 0, accumulator)
```
Comment (tcal-x) [inline proj/hps_accel/gateware/gen2/macc.py:156]: Is [0] the input end of the delay register, or the output end?

**Review by tcal-x — APPROVED** [review]
LGTM

File: proj/hps_accel/gateware/gen2/macc.py (hunk 1 above)
Comment (alanvgreen) [reply proj/hps_accel/gateware/gen2/macc.py:156]: Yeah, I'm not 100% happy with the API. Will probably make it less noisy in future.

Comment (alanvgreen) [reply proj/hps_accel/gateware/gen2/macc.py:156]: I made a better API in the next PR.

File: proj/hps_accel/gateware/gen2/utils.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Utility functions."""
+
+from nmigen import unsigned, Mux
+
+
+def unsigned_upto(maximum_value):
```
Comment (danc86) [inline proj/hps_accel/gateware/gen2/utils.py:20]: Hmm yes may reduce the likelihood of signal width mistakes :-)

File: proj/hps_accel/gateware/gen2/post_process.py [hunk 3]
```diff
@@ -282,7 +282,7 @@ def elab(self, m: Module):
 # Parameters for the ReadngProducer
 READING_PRODUCER_PARAMS = [
     ('depth', unsigned_upto(Constants.MAX_CHANNEL_DEPTH)),
-    ('repeats', unsigned(Constants.SYS_ARRAY_MAX_INPUTS)),
+    ('repeats', unsigned_upto(Constants.SYS_ARRAY_MAX_INPUTS)),
```
Comment (danc86) [inline proj/hps_accel/gateware/gen2/post_process.py:285]: :sweat_smile:

**Review by danc86 — APPROVED** [review]
Wow

File: proj/hps_accel/gateware/gen2/macc.py [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+    This block provides no flow control mechanisms. It is expected that
+    new inputs be provided on every cycle, and that output will be
+    processed when it is available. Avoiding flow control mechanisms
+    reduces the complexity and area use of this component, at the cost
+    of some additional external complexity.
+
+    The A input also carries 'first' and 'last signals' which control
+    accumulation.
+
+    This logic is intended to map to the Crosslink/NX SysDSP block.
```
[the remark this answers is not in the document]
Comment (alanvgreen) [reply proj/hps_accel/gateware/gen2/macc.py:43]: I find myself looking at the DSP usage count. There are likely better ways, which we will learn, in time.

**Pull request merged**